### PR TITLE
Migrate Mcp23xxx from .NET IoT to .NET nanoFramework

### PR DESCRIPTION
**Pull request by mork2020**
## Description

- migrated Mcp23xxx from .NET IoT to .NET nanoFramework   -   I2C devices only

Here are some file, which are not in the correct "place"
- SpanPinValuePair.cs --> where should this file go? System.Device.Gpio ? 
- PinVector32 --> where should this file go? System.Device.Gpio ?
- PinValuePairEnum.cs --> where should this file go? System.Device.Gpio ?
- GpioDriver.cs --> where should this file go? System.Device.Gpio ?

This file exist because of my error with symbol reference resolving (System.Device.Gpio.PinValuePair)
- PinValuePair.cs

## Motivation and Context

- new sensor is supported

## How Has This Been Tested?

- tested with an ESP32 and a MCP23017 breakout board 

## Types of changes

- [x] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [ ] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [ ] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [ ] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [ ] I have added new tests to cover my changes.

### CODE REVIEW

**Comment by Ellerbach** [comment]
> * SpanPinValuePair.cs --> where should this file go? System.Device.Gpio ?

All the Span files are automatically generated from Span<something> into SpanSomething. I don't think you need it as I think we've removed those from the GpioController. So align the functions on the nano System.Device.Gpio. Don't add those if not in ano. That will simplify the code.

> * PinVector32 --> where should this file go? System.Device.Gpio ?

No, all those should stay there. They should not be included in System.Device.Gpio at all. Especially if you align the API on it, I think a large part of those will just go away.

File: devices/Mcp23xxx/I2cAdapter.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+            }
+
+            /// <inheritdoc/>
+            public override void Write(byte registerAddress, SpanByte data)
+            {
+                SpanByte output = new byte[data.Length + 1];
+                output[0] = registerAddress;
+                // SpanByte slice is different to Span<Byte> slice in case that index argument (here 1) is equal to size of slice
+                if (data.Length > 0)
+                    data.CopyTo(output.Slice(1));       // do not override output[0]
```
Comment (Ellerbach) [inline devices/Mcp23xxx/I2cAdapter.cs:43]: please use the full braket { and }

File: devices/Mcp23xxx/Mcp23xxx.cs [hunk 2]
```diff
@@ -0,0 +1,572 @@
...
+            PinChangeEventHandler callback) => throw new NotImplementedException();
+
+        /// <inheritdoc/>
+        protected override void RemoveCallbackForPinValueChangedEvent(int pinNumber, PinChangeEventHandler callback) =>
+            throw new NotImplementedException();
+
+        /// <inheritdoc/>
+        protected override int ConvertPinNumberToLogicalNumberingScheme(int pinNumber) => pinNumber;
+
+        /*
```
Comment (Ellerbach) [inline devices/Mcp23xxx/Mcp23xxx.cs:562]: if you want to keep this one commented, please add a remark why

File: devices/Mcp23xxx/PinValuePairEnum.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
+﻿using System;
+using System.Collections;
+
+namespace Iot.Device.Mcp23xxx
+{
+    public class PinValuePairEnum : IEnumerator
```
Comment (Ellerbach) [inline devices/Mcp23xxx/PinValuePairEnum.cs:6]: if this class is public, please adjust it to add the intellisense comments ///
It has to be on everything public or internal

File: devices/Mcp23xxx/PinValuePairEnum.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    {
+        public Iot.Device.Mcp23xxx.PinValuePair[] _pair;
+        int currentIndex = -1;
+        public PinValuePairEnum(Iot.Device.Mcp23xxx.PinValuePair[] list)
```
Comment (Ellerbach) [inline devices/Mcp23xxx/PinValuePairEnum.cs:10]: Also please respect the coding guidelines with extra line between functions

File: devices/Mcp23xxx/PinValuePairEnum.cs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        {
+            _pair = list;
+        }
+        public bool MoveNext()
```
Comment (Ellerbach) [inline devices/Mcp23xxx/PinValuePairEnum.cs:14]: here as well missing /// and extra line

File: devices/Mcp23xxx/PinValuePairEnum.cs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        {
+            currentIndex++;
+            return (currentIndex < _pair.Length);
+        }
+        object IEnumerator.Current
```
Comment (Ellerbach) [inline devices/Mcp23xxx/PinValuePairEnum.cs:19]: here as well and missing an explicit visibility

File: devices/Mcp23xxx/PinValuePairEnum.cs [hunk 7]
```diff
@@ -0,0 +1,45 @@
+﻿using System;
+using System.Collections;
+
+namespace Iot.Device.Mcp23xxx
+{
+    public class PinValuePairEnum : IEnumerator
+    {
+        public Iot.Device.Mcp23xxx.PinValuePair[] _pair;
+        int currentIndex = -1;
```
Comment (Ellerbach) [inline devices/Mcp23xxx/PinValuePairEnum.cs:9]: we are using explicit visibility, I guess this one is private. Also for the naming, anything private variable should start with _

File: devices/Mcp23xxx/PinValuePairEnum.cs [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+            return (currentIndex < _pair.Length);
+        }
+        object IEnumerator.Current
+        {
+            get
+            {
+                return Current;
+            }
+        }
+        public Iot.Device.Mcp23xxx.PinValuePair Current
```
Comment (Ellerbach) [inline devices/Mcp23xxx/PinValuePairEnum.cs:26]: missing /// and extra line

File: devices/Mcp23xxx/PinValuePairEnum.cs [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+                {
+                    return _pair[currentIndex];
+                }
+                catch (IndexOutOfRangeException)
+                {
+                    throw new InvalidOperationException();
+                }
+            }
+        }
+        public void Reset()
```
Comment (Ellerbach) [inline devices/Mcp23xxx/PinValuePairEnum.cs:40]: missing /// and extra line

File: devices/Mcp23xxx/GpioDriver.cs [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        protected abstract PinValue Read(int pinNumber);
+
+        /// <summary>
+        /// Writes a value to a pin.
+        /// </summary>
+        /// <param name="pinNumber">The pin number in the driver's logical numbering scheme.</param>
+        /// <param name="value">The value to be written to the pin.</param>
+        protected abstract void Write(int pinNumber, PinValue value);
+
+
```
[the remark this answers is not in the document]
Comment (mork2020) [reply devices/Mcp23xxx/GpioDriver.cs:75]: correted

File: devices/Mcp23xxx/Mcp23xxx.cs (hunk 2 above)
Comment (mork2020) [reply devices/Mcp23xxx/Mcp23xxx.cs:562]: removed

File: devices/Mcp23xxx/I2cAdapter.cs (hunk 1 above)
Comment (mork2020) [reply devices/Mcp23xxx/I2cAdapter.cs:43]: corrected

File: devices/Mcp23xxx/PinValuePairEnum.cs (hunk 5 above)
Comment (mork2020) [reply devices/Mcp23xxx/PinValuePairEnum.cs:14]: file not necessary, removed from Mcp23xxx project

File: devices/Mcp23xxx/PinValuePairEnum.cs (hunk 7 above)
Comment (mork2020) [reply devices/Mcp23xxx/PinValuePairEnum.cs:9]: file not necessary, removed from Mcp23xxx project

File: devices/Mcp23xxx/PinValuePairEnum.cs (hunk 9 above)
Comment (mork2020) [reply devices/Mcp23xxx/PinValuePairEnum.cs:40]: file not necessary, removed from Mcp23xxx project

File: devices/Mcp23xxx/PinValuePairEnum.cs (hunk 8 above)
Comment (mork2020) [reply devices/Mcp23xxx/PinValuePairEnum.cs:26]: file not necessary, removed from Mcp23xxx project

File: devices/Mcp23xxx/PinValuePairEnum.cs (hunk 6 above)
Comment (mork2020) [reply devices/Mcp23xxx/PinValuePairEnum.cs:19]: file not necessary, removed from Mcp23xxx project

File: devices/Mcp23xxx/PinValuePairEnum.cs (hunk 3 above)
Comment (mork2020) [reply devices/Mcp23xxx/PinValuePairEnum.cs:6]: file not necessary, removed from Mcp23xxx project

**Comment by mork2020** [comment]
> > * SpanPinValuePair.cs --> where should this file go? System.Device.Gpio ?
> 
> All the Span files are automatically generated from Span into SpanSomething. I don't think you need it as I think we've removed those from the GpioController. So align the functions on the nano System.Device.Gpio. Don't add those if not in ano. That will simplify the code.

I removed the SpanPinValuePair.cs file from the Mcp23xxx project. How is the SpanPinValuePair.cs file generated? The automatic Azure build stops with error that the class is missing.

File: devices/Mcp23xxx/PinVector32.cs [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+            Values = values;
+        }
+
+        /// <summary>
+        /// Construct from a span of <see cref="PinValuePair"/>s.
+        /// </summary>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// Thrown if <paramref name="pinValues"/> contains negative pin numbers or pin numbers higher than 31.
+        /// </exception>
+        public PinVector32(SpanPinValuePair pinValues)
```
Comment (Ellerbach) [inline devices/Mcp23xxx/PinVector32.cs:42]: Here you have the SpanPinVlauePair. So you may want to simplify this as well

**Comment by Ellerbach** [comment]
> I removed the SpanPinValuePair.cs file from the Mcp23xxx project. How is the SpanPinValuePair.cs file generated? The automatic Azure build stops with error that the class is missing.

it's because you have one left in the code :-) See my comment on this

File: devices/Mcp23xxx/PinVector32.cs (hunk 11 above)
Comment (mork2020) [reply devices/Mcp23xxx/PinVector32.cs]: replaced SpanPinValuePair by PinValuePair[]

**Review by Ellerbach — APPROVED** [review]
Thanks a lot @mork2020 for this port! Looks great now!

**Pull request merged**